File: packages/py/src/qorechain_rdk/config/matrix.py
"""The settlement -> proof-system compatibility matrix enforced by the chain.

- ``optimistic`` -> ``fraud``
- ``zk`` -> ``snark`` | ``stark``
- ``based`` -> ``none``
- ``sovereign`` -> ``none``
"""

from __future__ import annotations

from ..enums import ProofSystem, SettlementParadigm

#: Settlement paradigm to the proof systems it accepts.
SETTLEMENT_PROOF_MATRIX: dict[SettlementParadigm, tuple[ProofSystem, ...]] = {
    SettlementParadigm.OPTIMISTIC: (ProofSystem.FRAUD,),
    SettlementParadigm.ZK: (ProofSystem.SNARK, ProofSystem.STARK),
    SettlementParadigm.BASED: (ProofSystem.NONE,),
    SettlementParadigm.SOVEREIGN: (ProofSystem.NONE,),
}


def valid_proof_systems(
    settlement: SettlementParadigm,
) -> tuple[ProofSystem, ...]:
    """Return the proof systems valid for a settlement paradigm."""
    return SETTLEMENT_PROOF_MATRIX[SettlementParadigm(settlement)]


def is_proof_compatible(
    settlement: SettlementParadigm, proof: ProofSystem
) -> bool:
    """Whether a proof system is compatible with a settlement paradigm."""
    return ProofSystem(proof) in SETTLEMENT_PROOF_MATRIX[SettlementParadigm(settlement)]


def requires_based_sequencer(settlement: SettlementParadigm) -> bool:
    """Whether a settlement paradigm requires the ``based`` sequencer mode."""
    return SettlementParadigm(settlement) == SettlementParadigm.BASED


__all__ = [
    "SETTLEMENT_PROOF_MATRIX",
    "valid_proof_systems",
    "is_proof_compatible",
    "requires_based_sequencer",
]

File: packages/jvm/src/main/java/io/github/qorechain/rdk/config/Matrix.java
package io.github.qorechain.rdk.config;

import io.github.qorechain.rdk.config.Enums.ProofSystem;
import io.github.qorechain.rdk.config.Enums.SettlementParadigm;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * The settlement → proof-system compatibility matrix enforced by the chain:
 *
 * <ul>
 *   <li>{@code optimistic} → {@code fraud}
 *   <li>{@code zk} → {@code snark} | {@code stark}
 *   <li>{@code based} → {@code none}
 *   <li>{@code sovereign} → {@code none}
 * </ul>
 */
public final class Matrix {
    private Matrix() {}

    /** The settlement → valid proof systems mapping. */
    public static final Map<SettlementParadigm, List<ProofSystem>> SETTLEMENT_PROOF_MATRIX;

    static {
        Map<SettlementParadigm, List<ProofSystem>> m = new EnumMap<>(SettlementParadigm.class);
        m.put(SettlementParadigm.OPTIMISTIC, List.of(ProofSystem.FRAUD));
        m.put(SettlementParadigm.ZK, List.of(ProofSystem.SNARK, ProofSystem.STARK));
        m.put(SettlementParadigm.BASED, List.of(ProofSystem.NONE));
        m.put(SettlementParadigm.SOVEREIGN, List.of(ProofSystem.NONE));
        SETTLEMENT_PROOF_MATRIX = Collections.unmodifiableMap(m);
    }

    /** Returns the proof systems valid for a settlement paradigm. */
    public static List<ProofSystem> validProofSystems(SettlementParadigm settlement) {
        List<ProofSystem> list = SETTLEMENT_PROOF_MATRIX.get(settlement);
        return list == null ? Collections.emptyList() : list;
    }

    /** Whether a proof system is compatible with a settlement paradigm. */
    public static boolean isProofCompatible(SettlementParadigm settlement, ProofSystem proof) {
        return validProofSystems(settlement).contains(proof);
    }

    /**
     * Whether a settlement paradigm requires the {@code based} sequencer mode. Only {@code based}
     * settlement carries this constraint.
     */
    public static boolean requiresBasedSequencer(SettlementParadigm settlement) {
        return settlement == SettlementParadigm.BASED;
    }
}
